### PR TITLE
Add timezone support to time_bucket

### LINKED ISSUE #414: Support for month, year and time zones in time_bucket
Time_bucket feature works great but why can't we use `SELECT time_bucket('1 month', date)...`

I would like to know how to get month bucket and even yearly buckets?

### PR DESCRIPTION
**Pull request by svenklemm**
This patch adds a new function time_bucket(period,timestamp,timezone)
which supports bucketing for arbitrary timezones.

Fixes #414

### CODE REVIEW

File: src/func_cache.c [hunk 1]
```diff
@@ -108,6 +109,22 @@ time_bucket_sort_transform(FuncExpr *func)
...
 }
 
+/*
+ * time_bucket with timezone will always have 5 args. For the sort
+ * optimization to apply all args need to be Const except timestamp.
+ */
+static Expr *
+time_bucket_tz_sort_transform(FuncExpr *func)
+{
+	Assert(list_length(func->args) == 5);
```
Comment (akuzm) [inline src/func_cache.c:119]: Looks like the tests are missing.

File: src/time_bucket.c [hunk 2]
```diff
@@ -281,6 +281,64 @@ ts_timestamptz_bucket(PG_FUNCTION_ARGS)
...
+	 */
+	if (PG_ARGISNULL(0) || PG_ARGISNULL(1) || PG_ARGISNULL(2))
+		PG_RETURN_NULL();
+
+	/* Convert to local timestamp according to timezone */
+	timestamp = DirectFunctionCall2(timestamptz_zone, tzname, timestamp);
+	if (have_offset)
+	{
+		/* Apply offset. */
+		timestamp = DirectFunctionCall2(timestamp_mi_interval, timestamp, PG_GETARG_DATUM(4));
```
Comment (erimatnor) [inline src/time_bucket.c:317]: Lots of test cases missing here according to codecov.

File: tsl/src/continuous_aggs/create.c [hunk 3]
```diff
@@ -817,51 +823,54 @@ caggtimebucket_validate(CAggTimebucketInfo *tbinfo, List *groupClause, List *tar
...
-				{
-					/*
-					 * Custom origin specified. This is always treated as
-					 * a variable-sized bucket case.
-					 */
-					tbinfo->bucket_width = BUCKET_WIDTH_VARIABLE;
+			}
+
+			if (list_length(fe->args) >= 5)
+				check_time_bucket_argument(lfourth(fe->args), "fifth");
```
Comment (akuzm) [inline tsl/src/continuous_aggs/create.c:829]: Should be `lfifth` I guess. There several other lines not covered by the tests, makes sense to add coverage for them as well.

**Review by akuzm — APPROVED** [review]

File: tsl/src/continuous_aggs/create.c [hunk 4]
```diff
@@ -184,6 +185,7 @@ typedef struct CAggTimebucketInfo
 	Interval *interval;			  /* stores the interval, NULL if not specified */
 	const char *timezone;		  /* the name of the timezone, NULL if not specified */
 
+	FuncExpr *bucket_func;
```
Comment (konskov) [inline tsl/src/continuous_aggs/create.c:188]: perhaps add a comment here for this field as well?

**Review by konskov — APPROVED** [review]

File: tsl/src/continuous_aggs/create.c (hunk 4 above)
Comment (svenklemm) [reply tsl/src/continuous_aggs/create.c]: done

**Pull request merged**